vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,675 @@
+  // be the config with the least exclusion. Note that all configs will run to budget
+  // before priorities and budget are reset.
+  float get_priority(int config_index)
+  {
+    float priority = 0.f;
+    for (auto ns : configs[config_index].exclusions) { priority -= ns_counter[ns]; }
+    return priority;
+  }
+
+  // This will generate configs with all combinations of current namespaces. These
+  // combinations will be held stored as 'exclusions' but can be used differently
+  // depending on oracle design.
+  void gen_configs(const multi_ex& ecs)
+  {
+    std::set<namespace_index> new_namespaces;
+    for (example* ex : ecs)
+    {
+      for (auto& ns : ex->indices)
+      {
+        ns_counter[ns]++;
+        if (ns_counter[ns] == 1) { new_namespaces.insert(ns); }
+      }
+    }
+    for (auto& ns : new_namespaces)
+    {
+      std::set<std::set<namespace_index>> new_exclusions;
+      for (auto& config : configs)
+      {
+        std::set<namespace_index> new_exclusion(config.second.exclusions);
+        new_exclusion.insert(ns);
+        new_exclusions.insert(new_exclusion);
+      }
+      for (auto& new_exclusion : new_exclusions)
+      {
+        int config_index = static_cast<int>(configs.size());
+        exclusion_config conf(starting_budget);
+        conf.exclusions = new_exclusion;
+        configs[config_index] = conf;
+        float priority = get_priority(config_index);
+        index_queue.push(std::make_pair(priority, config_index));
+      }
+    }
+  }
+
+  // This function is triggered when all sets of interactions generated by the oracle have been tried and
+  // reached their budget. It will then add inactive configs (stored in the config manager) to the queue
+  // 'index_queue' which can be used to swap out live configs as they run out of budget. This functionality
+  // may be better within the oracle, which could generate better priorities for different configs based
+  // on ns_counter (which is updated as each example is processed)
+  bool repopulate_index_queue()
+  {
+    std::set<int> live_indices;
+    for (size_t stride = 0; stride < max_live_configs; ++stride) { live_indices.insert(scores[stride].config_index); }
+    for (auto ind_config : configs)
+    {
+      int redo_index = ind_config.first;
+      if (live_indices.find(redo_index) == live_indices.end())
+      {
+        float priority = get_priority(redo_index);
+        index_queue.push(std::make_pair(priority, redo_index));
+      }
+    }
+    return !index_queue.empty();
+  }
+
+  // This function defines the logic update the live configs' budgets, and to swap out
+  // new configs when the budget runs out.
+  void update_live_configs()
+  {
+    for (size_t stride = 0; stride < max_live_configs; ++stride)
+    {
+      int config_index = scores[stride].config_index;
+      // Special case to initialize first champion to -q ::
+      if (config_index == -2)
```

Comment:
This special case can be removed if this logic is moved to the constructor of config_manager. This will allow us to remove the need for this -2 case. It could also be moved to the switch case of idle, when preparing the first experiment.

Reply:
This case is fixed by regenerating interactions whenever a new namespace is seen